pkg/github/orgreadwriter.go

```diff
@@ -0,0 +1,276 @@
+				users[v] = m
+			}
+		}
+		return resp, nil
+	}); err != nil {
+		return nil, err
+	}
+
+	members := make([]groupsync.Member, 0, len(users))
+	for _, user := range users {
+		members = append(members, &groupsync.UserMember{Usr: &groupsync.User{ID: user.GetLogin(), Attributes: user}})
+	}
+
+	return members, nil
+}
+
+// Descendants retrieve all users (children, recursively) of the GitHub org with the given ID.
+// For orgs, in practice this is the same as GetMembers.
+func (rw *OrgMembershipReadWriter) Descendants(ctx context.Context, groupID string) ([]*groupsync.User, error) {
+	logger := logging.FromContext(ctx)
+	logger.InfoContext(ctx, "fetching descendants for org", "org_id", groupID)
+	users, err := groupsync.Descendants(ctx, groupID, rw.GetMembers)
+	if err != nil {
+		return nil, fmt.Errorf("could not get descendants: %w", err)
+	}
+	return users, nil
+}
+
+// GetUser retrieves the GitHub user with the given ID. The ID is the GitHub user's login.
+func (rw *OrgMembershipReadWriter) GetUser(ctx context.Context, userID string) (*groupsync.User, error) {
+	user, err := rw.getGitHubUser(ctx, rw.client, userID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get user: %w", err)
+	}
+	return &groupsync.User{
+		ID:         user.GetLogin(),
+		Attributes: user,
+	}, nil
+}
+
+func (rw *OrgMembershipReadWriter) getGitHubUser(ctx context.Context, client *github.Client, userID string) (*github.User, error) {
+	if user, ok := rw.userCache.Lookup(userID); ok {
+		return user, nil
+	}
+	logger := logging.FromContext(ctx)
+	logger.InfoContext(ctx, "fetching user", "user_id", userID)
+	user, _, err := client.Users.Get(ctx, userID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch user %s: %w", userID, err)
+	}
+	rw.userCache.Set(userID, user)
+	return user, nil
+}
+
+// SetMembers replaces the members of the GitHub org with the given ID with the given members.
+// Any members of the GitHub org not found in the given members list will be removed.
+// Likewise, any members of the given list that are not currently members of the org will be added.
+func (rw *OrgMembershipReadWriter) SetMembers(ctx context.Context, groupID string, members []groupsync.Member) error {
+	orgID, err := strconv.ParseInt(groupID, 10, 64)
+	if err != nil {
+		return fmt.Errorf("could not parse orgID from groupID: %s", groupID)
+	}
+	client, err := rw.githubClientForOrg(ctx, orgID)
+	if err != nil {
+		return fmt.Errorf("could not create github client: %w", err)
+	}
+	org, err := rw.getGitHubOrg(ctx, client, orgID)
+	if err != nil {
+		return fmt.Errorf("failed to get org: %w", err)
+	}
+	orgName := org.GetName()
+	currentMembers, err := rw.GetMembers(ctx, groupID)
+	if err != nil {
+		return fmt.Errorf("could not get current members: %w", err)
+	}
+
+	// GitHub usernames are case-insensitive. So we should map each id
+	// to lower case before determining who to add and remove.
+	currentMemberIDs := toIDMap(currentMembers)
+	newMemberIDs := toIDMap(members)
```

Comment:
future improvement: I wonder if we should use user id instead of user name. I will create a github issue about this.